atv/decorators.py

```diff
@@ -0,0 +1,101 @@
+from functools import wraps
+
+from django.conf import settings
+from django.utils.translation import ugettext_lazy as _
+from rest_framework.exceptions import NotAuthenticated, PermissionDenied
+
+from atv.exceptions import MissingServiceAPIKey, ServiceNotIdentifiedError
+from services.enums import ServicePermissions
+from services.models import Service, ServiceAPIKey
+
+
+def _use_request_tests(*test_funcs):
+    def decorator(decorator_function):
+        @wraps(decorator_function)
+        def wrapper(function):
+            @wraps(function)
+            def request_tester(viewset, request, *args, **kwargs):
+                for test_func in test_funcs:
+                    test_func(request)
+
+                return function(viewset, request, *args, **kwargs)
+
+            return request_tester
+
+        return wrapper
+
+    return decorator
+
+
+def _require_service(request):
+    if not request.service:
+        raise ServiceNotIdentifiedError("No service identified")
+
+
+def _require_service_permission(permission_name):
+    def permission_checker(request):
+        if not request.user.is_superuser:
+            try:
+                _require_service(request)
+            except ServiceNotIdentifiedError:
+                raise PermissionDenied(
+                    _("You do not have permission to perform this action.")
+                )
+
+        if not request.user.has_perm(permission_name, request.service):
+            raise PermissionDenied(
+                _("You do not have permission to perform this action.")
+            )
+
+    return permission_checker
+
+
+def staff_required(required_permission=ServicePermissions.VIEW_DOCUMENTS):
+    """
+    Returns a decorator that checks if the user has staff permission on resources for the service
+    specified in the request. Required permission can be defined as an argument, defaults to 'view'.
+    """
+
+    if not isinstance(required_permission, ServicePermissions):
+        raise ValueError(
+            f"Invalid required_permission given as argument: '{required_permission}'"
+        )
+
+    @_use_request_tests(_require_service_permission(required_permission.value))
```

Comment:
This wrapped permission checking magic takes some time to fully understand what is happening, but it seems to be somewhat the same that happens in open-city-profile, so it should be fine. :sweat_smile: :+1:

Reply:
Took me a while to get it as well 😵 but yeah, I basically copy-pasted and adapted them to our case.